lib/block_view.go

```diff
@@ -2936,24 +3023,95 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
 	return false
 }
 
-func _verifySignature(txn *MsgBitCloutTxn) error {
-	// Compute a hash of the transaction
+func (bav *UtxoView) _verifySignature(txn *MsgBitCloutTxn, blockHeight uint32) error {
+	// Compute a hash of the transaction.
 	txBytes, err := txn.ToBytes(true /*preSignature*/)
 	if err != nil {
 		return errors.Wrapf(err, "_verifySignature: Problem serializing txn without signature: ")
 	}
 	txHash := Sha256DoubleHash(txBytes)
-	// Convert the txn public key into a *btcec.PublicKey
-	txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+
+	// We make a local copy of the signature as it should be read-only.
+	sigCopy := append([]byte{}, txn.Signature...)
+	// Verify that the control byte is within the allowed range.
+	// DERControlByte means txn was signed with the owner's private key.
+	// Otherwise, txn was signed with a derived key and we will attempt
+	// to recover the public key from the signature.
+	sigByte := sigCopy[0]
+	if !(sigByte >= DERControlByte && sigByte <= SIGMaxByte) {
+		return fmt.Errorf("_verifySignature: Signature control byte outside of allowed range")
+	}
+
+	// Reset the control byte to the DER control byte. Verify signature encoding.
+	sigCopy[0] = DERControlByte
+	sig, err := btcec.ParseDERSignature(sigCopy, btcec.S256())
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing signature: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// We check for public key solution iter for this transaction. This
+	// allows us to recover public key from the signature.
+	// If the transaction was signed with the owner key, iter is equal to 0.
+	// If the transaction was signed with a derived key, iter is between 1-4.
+	iter := sigByte - DERControlByte
```

Comment:
Nit: To me `iter` keyword means `iterator` and I'm expecting a loop. Can we pick a different name for this variable?